src/audio/module_adapter/module/cadence.c

```diff
@@ -21,6 +21,19 @@ DECLARE_SOF_RT_UUID("cadence_codec", cadence_uuid, 0xd8218443, 0x5ff3, 0x4a4c,
 
 DECLARE_TR_CTX(cadence_tr, SOF_UUID(cadence_uuid), LOG_LEVEL_INFO);
 
+/* Define a macro for error checking from Cadence API functions */
+
+#define XA_CHECK_ERROR(dev, err, fmt_extra, ret_statement, ...) ({ \
+	if (err != LIB_NO_ERROR) { \
+		if (XA_ERROR_SEVERITY(err)) {\
+			comp_err(dev, "Cadence API error %x" fmt_extra, err, ##__VA_ARGS__); \
+			ret_statement; \
```

Comment:
I don't understand how this works. 

It looks like you replaced
```
#define LIB_IS_FATAL_ERROR(e) ((e) & XA_FATAL_ERROR)
```
with:

```
#define XA_ERROR_SEVERITY(code) (((code) & XA_FATAL_ERROR) != 0)
```

Reply:
LIB_IS_FATAL_ERROR and XA_ERROR_SEVERITY are similar.
We should use XA_ERROR_SEVERITY - this is from Cadence API - https://github.com/thesofproject/sof/blob/main/src/include/sof/audio/cadence/xa_error_standards.h#L56

> Also, I don't really like XA_CHECK_ERROR .

Can you elaborate on this?

Reply:
I can use LIB_IS_FATAL_ERROR instead of XA_ERROR_SEVERITY, in the end it doesn't truly matter so I can do that before my next push.

Reply:
Use can use XA_ERROR_SEVERITY instead of 

> LIB_IS_FATAL_ERROR and XA_ERROR_SEVERITY are similar. We should use XA_ERROR_SEVERITY - this is from Cadence API - https://github.com/thesofproject/sof/blob/main/src/include/sof/audio/cadence/xa_error_standards.h#L56
> 
> > Also, I don't really like XA_CHECK_ERROR .
> 
> Can you elaborate on this?

I can agree we can use XA_ERROR_SEVERITY instead of LIB_IS_FATAL_ERROR but the name XA_ERROR_SEVERITY isn't really very useful. You don't know exactly what it does by only just looking at the name.

With LIB_IS_FATAL_ERROR you know you are testing for a FATAL error.

As for XA_CHECK_ERROR. I don't like it because it can just get just "return" from the current function and you don't really expect macros to do that. 

Just look at the call:

```XA_CHECK_ERROR(dev, ret, " while applying parameter %d with value %d", return ret, param->id, *(int32_t *)param->data);
```

It does have a 'return ret' as a parameter.

Reply:
I guess having manual checks instead (e.g. if IS_FATAL_ERROR <do something> else if is an error <do something else>) as in having this macro unwound is an idea?

I guess I'll remove commented out code, try to change this and open this up for wider review.